### PR TITLE
Change split point calculation in KD-tree construction

### PR DESCRIPTION
**Pull request by andreasnoack**
This tries to mimic the splitting of the original Loess implementation which is used by R. The implementation is based on reverse enginerring of the behavior as the rules are only loosely described in the original paper. With the rules described in the comment we are able to match the splits of R.

When adding tests, I realized that the weight calculation in the local regression were off by a square root. They were computed as the diagonal elements of W in `inv(X'*W*X)*X'*W*y` but we applied them to X and y before computing the OLS estimates so the weights were squared.

The signatures have also been loosened to allow more element types. This made it easier to test with the cars dataset from R.

I've added a lot of `@debug` statements to made it easier to follow the KD-tree construction.

Update: The changes to the signatures fixes https://github.com/JuliaStats/Loess.jl/issues/48

### CODE REVIEW

**Comment by andreasnoack** [comment]
I forgot to mention that, with these changes, I was able to change one of the broken tests to a working test. The other broken test was changed to a `@test_throws` as the `span` is now too small. The reason for this is that I'm no longer using `ceil` when calculating the minimum bucket size.

File: src/kd.jl [hunk 1]
```diff
@@ -55,11 +56,14 @@ function KDTree(xs::AbstractMatrix{T},
 
     diam = diameter(bounds)
 
-    leaf_size_cutoff = ceil(Int, leaf_size_factor * n)
+    # leaf_size_cutoff = ceil(Int, leaf_size_factor * n)
+    leaf_size_cutoff = leaf_size_factor * n
```
Comment (devmotion) [inline src/kd.jl:60]: Are non-integer `leaf_size_cutoff` values useful in the algorithm below? I noticed it is compared with `length(perm)` but there surely an integer would be sufficient?

File: src/Loess.jl [hunk 2]
```diff
@@ -178,14 +183,7 @@ function predict(model::LoessModel{T}, zs::AbstractVector{T}) where T <: Abstrac
...
 
-function predict(model::LoessModel{T}, zs::AbstractMatrix{T}) where T <: AbstractFloat
-    ys = Array{T}(undef, size(zs, 1))
-    for i in 1:size(zs, 1)
-        # the vec() here is not necessary on 0.5 anymore
-        ys[i] = predict(model, vec(zs[i,:]))
-    end
-    ys
-end
+predict(model::LoessModel, zs::AbstractMatrix) = [predict(model, zs[i,:]) for i in 1:size(zs, 1)]
```
Comment (devmotion) [inline src/Loess.jl:186]: To be safe:
```suggestion
predict(model::LoessModel, zs::AbstractMatrix) = [predict(model, zs[i,:]) for i in axes(zs, 1)]
```
or maybe just
```suggestion
predict(model::LoessModel, zs::AbstractMatrix) = map(Base.Fix1(predict, model), eachrow(zs))
```
(uses views but maybe for rows the allocating version is faster?)

File: src/Loess.jl [hunk 3]
```diff
@@ -135,12 +141,11 @@ end
 # Returns:
 #   A length n' vector of predicted response values.
 #
-function predict(model::LoessModel{T}, z::T) where T <: AbstractFloat
-    predict(model, T[z])
+function predict(model::LoessModel, z::Number)
```
Comment (devmotion) [inline src/Loess.jl:144]: Do we expect non-`Real` inputs?

File: src/Loess.jl [hunk 4]
```diff
@@ -135,12 +141,11 @@ end
...
+    predict(model, [z])
```
Comment (devmotion) [inline src/Loess.jl:145]: Could we move the algorithm below for `zs::AbstractVector` to an internal function `_predict(::LoessModel, ::Union{AbstractVector,Tuple})` that can be called with a `Tuple` as well (or just change the signature of `predict` below but this seems a bit weird maybe)? From a quick look at the implementation it seems that maybe the only additional change needed would be to relax `traverse(::KDTree, ::AbstractVector)` in the same way.
```suggestion
    predict(model, (z,))
```

Comment (devmotion) [reply src/Loess.jl:145]: An alternative, of course, would be to implement the special scalar case separately but then code would be a bit duplicated.

File: src/Loess.jl [hunk 5]
```diff
@@ -135,12 +141,11 @@ end
...
 end
 
-
-function predict(model::LoessModel{T}, zs::AbstractVector{T}) where T <: AbstractFloat
+function predict(model::LoessModel, zs::AbstractVector)
```
Comment (devmotion) [inline src/Loess.jl:148]: Not a new problem, so possible better addressed separately: This function fails if `zs` does not use one-based indexing.

File: src/kd.jl [hunk 6]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
-        mid = (length(perm) + 1) ÷ 2
-        partialsort!(perm, mid, by=i -> xs[i, j])
-        med = xs[perm[mid], j]
-        mid1 = mid
-        mid2 = mid + 1
-    else
-        mid1 = length(perm) ÷ 2
+    # Find the "median" and partition
+    #
+    # The aim of the alogorihtm is to split the data recursively in two roughly equally sized
```
Comment (devmotion) [inline src/kd.jl:151]:
```suggestion
    # The aim of the algorithm is to split the data recursively in two roughly equally sized
```

File: src/kd.jl [hunk 7]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
+    # subsets. To do so, we'll use the median element of x[:, j] but there are a
+    # few corner cases that require some care: sets with an even number of elements
+    # doesn't have a unique median and ties. Below, we'll list the possibilities.
+    #
+    # - Odd number of elements and no ties, e.g. [1, 2, 3]: the median is
+    #   unambiguously 2 and split the data into [1, 2] and [3]
+    #
+    # - Even number of element and no ties, e.g. [1, 2, 3, 4]: we choose the left middle
+    #   value 2 as median and split the dato into [1, 2] and [3, 4]
```
Comment (devmotion) [inline src/kd.jl:160]:
```suggestion
    #   value 2 as median and split the data into [1, 2] and [3, 4]
```

File: src/kd.jl [hunk 8]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
+    #
+    # - Ties will cause a search to the change in value that divides the set most evenly.
+    #   E.g. [1, 2, 2, 3] uses 2 as median value to split the data into [1, 2, 2] and [3]
+    #   but [1, 1, 2, 2, 2] uses 1 to split into [1, 1] and [2, 2, 2] even though 1 is
+    #   not a proper median value. This avoids that the same value in two buckets.
```
Comment (devmotion) [inline src/kd.jl:165]:
```suggestion
    #   not a proper median value. This avoids that the same value is in two buckets.
```

File: src/kd.jl [hunk 9]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
+    #
+    # The details here are reversed engineered from the C/Fortran implementation wrapped
+    # by R and also distribtued on NETLIB.
+    mid = (length(perm) + 1) ÷ 2
+    @debug "Candidate median index and median value" mid xs[perm[mid], j]
```
Comment (devmotion) [inline src/kd.jl:170]: Note that again this assumes that `perm` and `xs` use 1-based indexing. Maybe just add some check for it using `Base.require_one...` for now?

File: src/kd.jl [hunk 10]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
-        med = (xs[perm[mid1], j] + xs[perm[mid2], j]) / 2
+        if mid1 < 1
+            @debug "mid1 is zero. All elements are identical. Creating vertex and then two leaves" mid1 length(perm) xs[perm[mid], j]
+            offset = mid1 = 0
+            mid2 = length(perm) + 1
+            break
+        end
+        if mid2 > length(perm)
+            @debug "mid2 is out of bounds. Continuing with negative offset" mid2 length(perm) offset
+            offset = -offset + (offset <= 0)
```
Comment (devmotion) [inline src/kd.jl:185]: Can you explain this line, maybe with an example?

File: src/kd.jl [hunk 11]
```diff
@@ -141,19 +146,56 @@ function build_kdtree(xs::AbstractMatrix{T},
...
+            continue
+        end
+        p12 = partialsort!(perm, mid1:mid2, by = i -> xs[i, j])
+        if xs[p12[1], j] == xs[p12[2], j]
+            @debug "tie! Adjusting offset" xs[p12[1], j] xs[p12[2], j] offset
+            offset = -offset + (offset <= 0)
```
Comment (devmotion) [inline src/kd.jl:191]: Same here, maybe a short example or explanation?

File: src/Loess.jl [hunk 12]
```diff
@@ -89,8 +95,8 @@ function loess(xs::AbstractMatrix{T}, ys::AbstractVector{T};
```
[the remark this answers is not in the document]
Comment (andreasnoack) [reply src/Loess.jl:95]: Indeed. Let's include the fix here.

File: src/Loess.jl (hunk 3 above)
Comment (andreasnoack) [reply src/Loess.jl:144]: No. It might work but let's restrict it until there is a demand and it has been tested.

File: src/Loess.jl (hunk 4 above)
Comment (andreasnoack) [reply src/Loess.jl:145]: I plan to make optimizations along those lines in a follow-up PR.

File: src/kd.jl (hunk 1 above)
Comment (andreasnoack) [reply src/kd.jl:60]: No. Changing it to `floor` should suffice here.

**Comment by andreasnoack** [comment]
Thanks for the comments. I believe that I've now addressed all of them so please have another look.

**Review by devmotion — APPROVED** [review]
Looks good to me :slightly_smiling_face:

**Pull request merged**